### PR TITLE
Object detection support dataframe, dict, list, str input for .fit(), .predict(), .evaluate()

### PR DESCRIPTION
**Pull request by yongxinw**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 1]
```diff
@@ -682,16 +683,33 @@ def fit(
...
-                self.detection_anno_train = tuning_data
-                tuning_data = from_coco_or_voc(tuning_data, "val")
-                if max_num_tuning_data is not None:
-                    if len(tuning_data) > max_num_tuning_data:
-                        tuning_data = tuning_data.sample(
-                            n=max_num_tuning_data, replace=False, random_state=seed
-                        ).reset_index(drop=True)
+            if isinstance(train_data, str):
+                self._detection_anno_train = train_data
+                train_data = from_coco_or_voc(train_data, "train")  # TODO: Refactor to use convert_data_to_df
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/predictor.py:688]: Need an example or tutorial to visualize what the dataframe looks like, because one image path can have multiple bboxes with coordinates. Users with no experience in object detection may not know what the format looks like. In addition, because we support dataframe now, users can directly prepare data in dataframe format without converting to coco/voc format first.

**Review by bryanyzhu — APPROVED** [review]
LGTM, but needs approval from @FANGAreNotGnu to confirm, thank you.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 1 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/predictor.py:688]: Yea totally agree. I think it is necessary to do such a tutorial for `.fit()` with df to make sure they are passing the right format.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 2]
```diff
@@ -682,16 +683,33 @@ def fit(
...
+                    tuning_data = autogluon.multimodal.utils.object_detection.convert_data_to_df(tuning_data)
+                    if max_num_tuning_data is not None:
+                        if len(tuning_data) > max_num_tuning_data:
+                            tuning_data = tuning_data.sample(
+                                n=max_num_tuning_data, replace=False, random_state=seed
+                            ).reset_index(drop=True)
+            else:
+                raise TypeError(
+                    f"Expected train_data to have type str or pd.DataFrame, but got type: {type(train_data)}"
+                )
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:712]: Can we wrap this part into a util function of `object_detection.py`? Let's simplify `predictor.py`.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
 from packaging import version
 from torch import nn
 
+import autogluon.multimodal.utils.object_detection
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:29]: This import style is different from others. Could we do `from .utils import convert_data_to_df`? Maybe also consider changing `convert_data_to_df` to `object_detection_data_to_df` to let the name more specific.

File: multimodal/tests/unittests/others/test_object_detection.py [hunk 4]
```diff
@@ -88,6 +88,48 @@ def test_mmdet_object_detection_fit_then_evaluate_coco(checkpoint_name):
...
+    predictor = MultiModalPredictor(
+        hyperparameters={
+            "model.mmdet_image.checkpoint_name": checkpoint_name,
+            "env.num_gpus": 1,  # currently mmdet only support single gpu inference
+        },
+        problem_type="object_detection",
+    )
+
+    pred = predictor.predict(mmdet_image_name)  # test batch inference
+    assert len(pred) == 1  # test data size is 100
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_object_detection.py:130]: Could we combine the two tests? Most code is shared.

File: multimodal/tests/unittests/others/test_object_detection.py [hunk 5]
```diff
@@ -290,3 +332,71 @@ def test_mmdet_object_detection_fit_then_inference_coco(checkpoint_name):
...
+        train_df,
+        hyperparameters={
+            "optimization.learning_rate": 2e-4,
+            "env.per_gpu_batch_size": 2,
+        },
+        time_limit=30,
+    )
+
+    preds = predictor.predict(data=test_df)
+    results = predictor.evaluate(data=test_df)
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_object_detection.py:402]: Ditto. Consider combining the two tests.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 6]
```diff
@@ -1863,14 +1881,25 @@ def evaluate(
...
+                return evaluate_coco(
+                    predictor=self,
+                    anno_file_or_df=data,
+                    metrics=metrics,
+                    return_pred=return_pred,
+                    seed=seed,
+                    eval_tool=eval_tool,
+                )
+            else:
+                data = autogluon.multimodal.utils.object_detection.convert_data_to_df(data)
```
Comment (FANGAreNotGnu) [inline multimodal/src/autogluon/multimodal/predictor.py:1894]: shall we do `from .util import convert_data_to_df` for this?

File: multimodal/src/autogluon/multimodal/utils/object_detection.py [hunk 7]
```diff
@@ -16,7 +16,99 @@
...
+    ----------
+    data
+        dataframe holding the data for object detection
+    Returns
+    -------
+
+    """
+    if "image" not in data:
+        raise ValueError(f"column 'image' not found in data column names: {data.columns.to_list()}")
+    if "rois" not in data:
```
Comment (FANGAreNotGnu) [inline multimodal/src/autogluon/multimodal/utils/object_detection.py:63]: We need only one between rois and label. And we could make a copy internally.

File: multimodal/src/autogluon/multimodal/utils/object_detection.py [hunk 8]
```diff
@@ -16,7 +16,99 @@
...
+        raise ValueError(f"column 'rois' not found in data column names: {data.columns.to_list()}")
+    if "label" not in data:
+        raise ValueError(f"column 'label' not found in data column names: {data.columns.to_list()}")
+    assert data.shape[0] > 0, "data frame is empty"
+
+
+def from_str(data: str) -> pd.DataFrame:
```
Comment (FANGAreNotGnu) [inline multimodal/src/autogluon/multimodal/utils/object_detection.py:70]: Xingjian mentioned that we should consider a json str. We can mark it as TODO and discuss later.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 2 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/predictor.py:712]: yup

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/predictor.py:29]: Yea. This was because of one of the previous conversation with xingjian where he mentioned to use this kind of import instead of from `.utils import convert_data_to_df`. I can change it now

File: multimodal/tests/unittests/others/test_object_detection.py (hunk 5 above)
Comment (yongxinw) [reply multimodal/tests/unittests/others/test_object_detection.py:402]: yea that makes sense

File: multimodal/src/autogluon/multimodal/utils/object_detection.py (hunk 8 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/utils/object_detection.py:70]: Yea definitely

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 9]
```diff
@@ -27,6 +27,11 @@
 
 from autogluon.common.utils.log_utils import set_logger_verbosity, verbosity2loglevel
 from autogluon.multimodal.utils import save_result_df
+from autogluon.multimodal.models.fusion import AbstractMultimodalFusionModel
+from autogluon.multimodal.models.huggingface_text import HFAutoModelForTextPrediction
+from autogluon.multimodal.models.mmdet_image import MMDetAutoModelForObjectDetection
+from autogluon.multimodal.models.timm_image import TimmAutoModelForImagePrediction
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:33]: It seems that these model classes are not used.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 10]
```diff
@@ -27,6 +27,11 @@
...
+from autogluon.multimodal.utils import object_detection_data_to_df, save_result_df, setup_detection_train_tuning_data
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:34]: Better to put the three functions inside the `from .utils import` below.

**Review by zhiqiangdon — APPROVED** [review]
LGTM!

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 9 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/predictor.py]: Yup they are deleted now
